### PR TITLE
fix(engine): don't mutate state when checking for timed out jobs

### LINKED ISSUE #12797: Don't mutate state through `JobTimeoutTrigger`
We need to make sure that `JobTimeoutTrigger` does not have the ability to mutate state.

`JobTimeoutTrigger.scheduleDeactivateTimedOutJobsTask` is able to mutate state while it should not be allowed to do so. The state may only be mutated by event appliers, see 
- https://github.com/camunda/zeebe/issues/6171

Once this is resolved, we could add an arch unit test to verify that only Event Appliers are using the mutable state.

This is a blocker for:
- #12302

### PR DESCRIPTION
**Pull request by lenaschoenburg**
This PR closes #12797, ensuring that job deadlines are reliably cleaned up by event appliers. This is necessary so that the job timeout checker does not have to mutate state.

With 91da1e89a9e39b29d6ae9f1cbf6a5f1ca77ce271, we fix the job yield and recur processors to produce events with the current job state. Previously, these propagated the job record from the command into new follow-up events which may overwrite already persisted changes to a job. 
With 330ce0b1980a609a978a9680f8a037070669090a, it is now clear when deadlines are set and removed. This follows the [rules we discussed](https://github.com/camunda/zeebe/issues/12797#issuecomment-1573731298), essentially adding a deadline on activation and removing whe the job is no longer activated or deleted.
Finally, cfdef6d037b78f5c78505049aa74bab00260035f is removing the illegal state modification from the checker. With the previous changes we can be sure that deadlines are cleaned up reliably and don't need to do redundant cleanup in the checker anymore.

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java [hunk 1]
```diff
@@ -234,8 +230,11 @@ private void updateJob(final long key, final JobRecord updatedValue, final State
       makeJobActivatable(type, key);
     }
 
-    if (deadline > 0) {
-      removeJobDeadline(deadline);
+    if (newState != State.ACTIVATED) {
+      // This only works because none of the events actually remove the deadline from the job record.
+      // If, say on job failure, the deadline is removed or reset to 0, then we would need to look at
+      // the current state of the job to determine what deadline to remove.
+      removeJobDeadline(key, updatedValue.getDeadline());
```
Comment (oleschoenburg) [inline engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:237]: :question: I'm not sure about this assumption: "none of the events actually remove the deadline from the job record"

File: engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeOutProcessor.java [hunk 2]
```diff
@@ -44,32 +45,29 @@ public JobTimeOutProcessor(
...
-        case FAILED:
-          textState = "it is marked as failed";
-          break;
-        default:
-          textState = "no such job was found";
-          break;
-      }
+      final var reason =
+          switch (state) {
+            case ACTIVATED -> "it has not timed out";
```
Comment (oleschoenburg) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeOutProcessor.java:63]: :upside_down_face: A bit inelegant, would be much nicer with switch guards, coming in Java 21.

**Review by korthout — APPROVED** [review]
Great job @oleschoenburg 👏 

👍 I love the usage of switch expression for the rejection reason! It reads like a breeze and highlights the previously missing cases that were wrongfully caught by the default case. And, as you say, it can be improved even further when we're on Java 21.

👍 I love the change to a BiPredicate as a better-defined type

👍 I love the addition of `addJobDeadline` to improve the DbJobSate code

👍 LGTM

bors merge

**Comment by korthout** [comment]
@oleschoenburg Maybe we should backport this 🤔

**Pull request merged**